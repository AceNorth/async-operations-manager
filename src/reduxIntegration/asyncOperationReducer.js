import _ from 'lodash';

import { getStateForOperationAfterStep } from '../asyncOperationManagerUtils';

import {
  ASYNC_OPERATION_STEPS,
} from '../constants';

function asyncOperationReducer(state = {}, action) {
  if (_.includes(ASYNC_OPERATION_STEPS, action.operationStep) && action.descriptorId) {
    return {
      ...state,
      ...getStateForOperationAfterStep(state, action.operationStep, action.descriptorId, action),
    };
  }
  return state;
}


export default asyncOperationReducer;
