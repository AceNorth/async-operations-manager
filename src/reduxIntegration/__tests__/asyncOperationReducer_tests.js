/* eslint-env jest */
import { expect } from 'chai';

import asyncOperationReducer from '../asyncOperationReducer';
import {
  clearAsyncOperationsManagerState,
  registerAsyncOperationDescriptors,
} from '../../asyncOperationManagerUtils';

describe('asyncOperationReducer', () => {
  let state;

  beforeEach(() => {
    state = {};
    clearAsyncOperationsManagerState();
    const dateNowStub = jest.fn(() => 1530518207007);
    global.Date.now = dateNowStub;
  });

  it('should return the initial state', () => {
    expect(asyncOperationReducer(undefined, {})).to.deep.equal(state);
  });
  it('should handle BEGIN_ASYNC_OPERATION for write action', () => {
    const action = {
      descriptorId: 'UPDATE_PERSON_DATA',
      operationStep: 'BEGIN_ASYNC_OPERATION',
      type: 'AOM//BEGIN__UPDATE_PERSON_DATA',
      actionType: 'UPDATE_PERSON_DATA',
      personId: 111,
    };

    registerAsyncOperationDescriptors({
      descriptorId: 'UPDATE_PERSON_DATA',
      requiredParams: ['personId'],
      operationType: 'WRITE',
    });

    expect(asyncOperationReducer(state, action)).to.have.all.keys('UPDATE_PERSON_DATA_111');
    expect(asyncOperationReducer(state, action)).to.matchSnapshot('well formed state for begun write async operation');
  });
  it('should handle RESOLVE_ASYNC_OPERATION for write action', () => {
    const action = {
      descriptorId: 'UPDATE_PERSON_DATA',
      operationStep: 'RESOLVE_ASYNC_OPERATION',
      type: 'AOM//RESOLVE__UPDATE_PERSON_DATA',
      actionType: 'UPDATE_PERSON_DATA',
      personId: 111,
    };

    registerAsyncOperationDescriptors({
      descriptorId: 'UPDATE_PERSON_DATA',
      requiredParams: ['personId'],
      operationType: 'WRITE',
    });

    expect(asyncOperationReducer(state, action)).to.have.all.keys('UPDATE_PERSON_DATA_111');
    expect(asyncOperationReducer(state, action)).to.matchSnapshot('well formed state for resolved write async operation');
  });
  it('should handle REJECT_ASYNC_OPERATION for write action', () => {
    const action = {
      descriptorId: 'UPDATE_PERSON_DATA',
      operationStep: 'REJECT_ASYNC_OPERATION',
      type: 'AOM//REJECT__UPDATE_PERSON_DATA',
      actionType: 'UPDATE_PERSON_DATA',
      personId: 111,
    };

    registerAsyncOperationDescriptors({
      descriptorId: 'UPDATE_PERSON_DATA',
      requiredParams: ['personId'],
      operationType: 'WRITE',
    });

    expect(asyncOperationReducer(state, action)).to.have.all.keys('UPDATE_PERSON_DATA_111');
    expect(asyncOperationReducer(state, action)).to.matchSnapshot('well formed state for rejected read async operation');
  });
  it('should handle BEGIN_ASYNC_OPERATION for read action', () => {
    const action = {
      descriptorId: 'FETCH_PERSON_DATA',
      operationStep: 'BEGIN_ASYNC_OPERATION',
      type: 'AOM//BEGIN__FETCH_PERSON_DATA',
      actionType: 'FETCH_PERSON_DATA',
      personId: 111,
    };

    registerAsyncOperationDescriptors({
      descriptorId: 'FETCH_PERSON_DATA',
      requiredParams: ['personId'],
      operationType: 'READ',
    });

    expect(asyncOperationReducer(state, action)).to.have.all.keys('FETCH_PERSON_DATA_111');
    expect(asyncOperationReducer(state, action)).to.matchSnapshot('well formed state for begun read async operation');
  });
  it('should handle RESOLVE_ASYNC_OPERATION for read action', () => {
    const action = {
      descriptorId: 'FETCH_PERSON_DATA',
      operationStep: 'RESOLVE_ASYNC_OPERATION',
      type: 'AOM//RESOLVE__FETCH_PERSON_DATA',
      actionType: 'FETCH_PERSON_DATA',
      personId: 111,
    };

    registerAsyncOperationDescriptors({
      descriptorId: 'FETCH_PERSON_DATA',
      requiredParams: ['personId'],
      operationType: 'READ',
    });

    expect(asyncOperationReducer(state, action)).to.have.all.keys('FETCH_PERSON_DATA_111');
    expect(asyncOperationReducer(state, action)).to.matchSnapshot('well formed state for resolved read async operation');
  });
  it('should handle REJECT_ASYNC_OPERATION for write action', () => {
    const action = {
      descriptorId: 'FETCH_PERSON_DATA',
      operationStep: 'REJECT_ASYNC_OPERATION',
      type: 'AOM//REJECT__FETCH_PERSON_DATA',
      actionType: 'FETCH_PERSON_DATA',
      personId: 111,
    };

    registerAsyncOperationDescriptors({
      descriptorId: 'FETCH_PERSON_DATA',
      requiredParams: ['personId'],
      operationType: 'WRITE',
    });

    expect(asyncOperationReducer(state, action)).to.have.all.keys('FETCH_PERSON_DATA_111');
    expect(asyncOperationReducer(state, action)).to.matchSnapshot('well formed state for rejected write async operation');
  });
});
