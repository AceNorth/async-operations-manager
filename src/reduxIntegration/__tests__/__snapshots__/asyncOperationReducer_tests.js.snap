// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`asyncOperationReducer should handle BEGIN_ASYNC_OPERATION for read action: well formed state for begun read async operation 1`] = `
Object {
  "FETCH_PERSON_DATA_111": Object {
    "dataStatus": "ABSENT",
    "descriptorId": "FETCH_PERSON_DATA",
    "fetchStatus": "PENDING",
    "lastDataStatusTime": 0,
    "lastFetchStatusTime": 1530518207007,
    "message": null,
    "personId": 111,
  },
}
`;

exports[`asyncOperationReducer should handle BEGIN_ASYNC_OPERATION for write action: well formed state for begun write async operation 1`] = `
Object {
  "UPDATE_PERSON_DATA_111": Object {
    "descriptorId": "UPDATE_PERSON_DATA",
    "fetchStatus": "PENDING",
    "lastFetchStatusTime": 1530518207007,
    "message": null,
    "personId": 111,
  },
}
`;

exports[`asyncOperationReducer should handle REJECT_ASYNC_OPERATION for write action: well formed state for rejected read async operation 1`] = `
Object {
  "UPDATE_PERSON_DATA_111": Object {
    "actionType": "UPDATE_PERSON_DATA",
    "descriptorId": "UPDATE_PERSON_DATA",
    "fetchStatus": "FAILED",
    "lastFetchStatusTime": 1530518207007,
    "message": null,
    "operationStep": "REJECT_ASYNC_OPERATION",
    "personId": 111,
    "type": "AOM//REJECT__UPDATE_PERSON_DATA",
  },
}
`;

exports[`asyncOperationReducer should handle REJECT_ASYNC_OPERATION for write action: well formed state for rejected write async operation 1`] = `
Object {
  "FETCH_PERSON_DATA_111": Object {
    "actionType": "FETCH_PERSON_DATA",
    "descriptorId": "FETCH_PERSON_DATA",
    "fetchStatus": "FAILED",
    "lastFetchStatusTime": 1530518207007,
    "message": null,
    "operationStep": "REJECT_ASYNC_OPERATION",
    "personId": 111,
    "type": "AOM//REJECT__FETCH_PERSON_DATA",
  },
}
`;

exports[`asyncOperationReducer should handle RESOLVE_ASYNC_OPERATION for read action: well formed state for resolved read async operation 1`] = `
Object {
  "FETCH_PERSON_DATA_111": Object {
    "actionType": "FETCH_PERSON_DATA",
    "dataStatus": "PRESENT",
    "descriptorId": "FETCH_PERSON_DATA",
    "fetchStatus": "SUCCESSFUL",
    "lastDataStatusTime": 1530518207007,
    "lastFetchFailed": false,
    "lastFetchStatusTime": 1530518207007,
    "message": null,
    "operationStep": "RESOLVE_ASYNC_OPERATION",
    "personId": 111,
    "type": "AOM//RESOLVE__FETCH_PERSON_DATA",
  },
}
`;

exports[`asyncOperationReducer should handle RESOLVE_ASYNC_OPERATION for write action: well formed state for resolved write async operation 1`] = `
Object {
  "UPDATE_PERSON_DATA_111": Object {
    "actionType": "UPDATE_PERSON_DATA",
    "descriptorId": "UPDATE_PERSON_DATA",
    "fetchStatus": "SUCCESSFUL",
    "lastFetchStatusTime": 1530518207007,
    "message": null,
    "operationStep": "RESOLVE_ASYNC_OPERATION",
    "personId": 111,
    "type": "AOM//RESOLVE__UPDATE_PERSON_DATA",
  },
}
`;
